setup.py

```diff
@@ -57,20 +57,30 @@ def package_data(pkg, roots):
     ],
     python_requires='>=3.11',
     install_requires=[
-        'apscheduler<3.8',
-        'google-api-python-client<1.8',
-        'paramiko',
-        'python-heatclient<2',
-        'python-keystoneclient<3.22',
-        'python-novaclient<16',
-        'tenacity<8',
+        'apscheduler<4',
+        'cliff<=4.7',
+        'dogpile.cache<=1.3',
+        'google-api-python-client<=2.149.0',
+        'keystoneauth1<=5.8',
+        'openstacksdk<=4.1',
+        'osc-lib<=3.1',
+        'os-client-config<=2.1',
+        'oslo.serialization<=5.5',
+        'oslo.utils<=7.4',
+        'oslo.config<=9.7.1',
+        'paramiko>=3.4.0',
```

Comment:
Just checking: is there a specific reason why we're defining an upper bound for everything else, but only a lower bound for paramiko?

Reply:
The lower bound we set for vulnerability fix in https://github.com/hastexo/hastexo-xblock/pull/287/commits/9d477f909c3dcf3edeccb4b3f4637d407a893b7e
I am not sure why we don't have an upper bound in this case.

Reply:
Okay, in that case let's add `<4`.

Reply:
Ack, done.